pkg/pipeline/pipeline.go

```diff
@@ -149,31 +156,81 @@ func NewPipeline() (*Pipeline, error) {
 		Writer:       writer,
 	}
 
+	// TODO: all this apply* functions can be removed if we change the Ingester, Transformer, etc...
+	// interfaces and use them directly here
+	ingests := node.AsInit(p.Ingester.Ingest)
+	decodes := node.AsMiddle(p.applyDecode)
+	transforms := node.AsMiddle(p.applyTransforms)
+	writes := node.AsTerminal(p.applyWrite)
+	extracts := node.AsMiddle(p.applyExtract)
+	encodes := node.AsTerminal(p.applyEncode)
+
```

Comment:
At present, @eranra uses the same transformed data to be fed into `write` and `extract` (which then feeds into `encode`). We probably want to combine `write` and `encode` to something like `export`. We may need to maintain the notion of `decode` since it takes a different format for its input than the other types of stages.

Reply:
I see the point. I think it's an interesting discussion.

Reply:
@mariomac we talked about a wrapper function that will allow using stages that do have output as `AsTerminal` phases. So that function will just drop the output of the phase so it will fit when needed. 

An example is the Prometheus encode phase ... it does have output and in some cases, for example, debug you want to push the output into another stdout write phase but for other cases, this should be a termination phase ... so with such wrapping, we can support both.

Reply:
Now I understand better this proposal after another comment from @KalmanMeth .

We can do two things:
- Change the "write_stdout" interface to work as middle node: receives an entry, submits it, forwards it.

<img width="948" alt="image" src="https://user-images.githubusercontent.com/939550/156531945-978fe258-0ff2-4232-b326-9ce739785c86.png">

- Don't do nothing by now. Just consider that you can currently attach terminal nodes to any stage of the pipeline:
 
<img width="789" alt="image" src="https://user-images.githubusercontent.com/939550/156531756-91ed12ec-e017-43bd-9ab3-10edc4dc52d9.png">